module/state_synchronization/indexer/in_memory_indexer.go

```diff
@@ -0,0 +1,221 @@
+		Msg("indexer initialized")
+
+	return indexer
+}
+
+// RegisterValue retrieves register values by the register IDs at the provided block height.
+// If a register is not found it will return a nil value and not an error.
+// Expected errors:
+// - storage.ErrHeightNotIndexed if the given height was not indexed yet or lower than the first indexed height.
+func (i *InMemoryIndexer) RegisterValue(ID flow.RegisterID, height uint64) (flow.RegisterValue, error) {
+	value, err := i.registers.Get(ID, height)
+	if err != nil {
+		// only return an error if the error doesn't match the not found error, since we have
+		// to gracefully handle not found values and instead assign nil, that is because the script executor
+		// expects that behavior
+		if errors.Is(err, storage.ErrNotFound) {
+			return nil, nil
+		}
+
+		return nil, err
+	}
+
+	return value, nil
+}
+
+// IndexBlockData indexes all execution block data.
+// Unlike the original implementation, this doesn't use batches as the in-memory
+// caches don't support batched writes.
+func (i *InMemoryIndexer) IndexBlockData(data *execution_data.BlockExecutionDataEntity) error {
+	log := i.log.With().Hex("block_id", logging.ID(data.BlockID)).Logger()
+	log.Debug().Msgf("indexing block data")
+
+	start := time.Now()
+
+	// Initialize collection containers
+	events := make([]flow.Event, 0)
+	results := make([]flow.LightTransactionResult, 0)
+	payloads := make(map[ledger.Path]*ledger.Payload)
+	indexedCollections := 0
+
+	// Process all chunk data in a single pass
+	for idx, chunk := range data.ChunkExecutionDatas {
+		// Collect events
+		events = append(events, chunk.Events...)
+
+		// Collect transaction results
+		results = append(results, chunk.TransactionResults...)
+
+		// Process collections (except system chunk)
+		if idx < len(data.ChunkExecutionDatas)-1 {
+			if err := i.handleCollection(chunk.Collection); err != nil {
+				return fmt.Errorf("could not handle collection: %w", err)
+			}
+			indexedCollections++
+		}
+
+		// Process register updates
+		if chunk.TrieUpdate != nil {
+			// Verify trie update integrity
+			if len(chunk.TrieUpdate.Paths) != len(chunk.TrieUpdate.Payloads) {
+				return fmt.Errorf("update paths length is %d and payloads length is %d and they don't match",
+					len(chunk.TrieUpdate.Paths), len(chunk.TrieUpdate.Payloads))
+			}
+
+			// Collect payloads (last one for a path wins)
+			for i, path := range chunk.TrieUpdate.Paths {
+				payloads[path] = chunk.TrieUpdate.Payloads[i]
+			}
+		}
+	}
+
+	// Store events
+	if err := i.events.Store(data.BlockID, []flow.EventsList{events}); err != nil {
+		return fmt.Errorf("could not index events: %w", err)
+	}
+
+	// Store transaction results
+	if err := i.results.Store(data.BlockID, results); err != nil {
+		return fmt.Errorf("could not index transaction results: %w", err)
+	}
+
+	// Store registers
+	if err := i.indexRegisters(payloads, i.registers.LatestHeight()); err != nil {
```

Comment:
I don't think we should rely on the storage to have preset the latest height before the data is stored. Instead, let's just pass the block header as a constructor argument and use that